src/main/java/org/elasticsearch/index/snapshots/blobstore/BlobStoreIndexShardSnapshots.java

```diff
@@ -31,19 +43,66 @@
  * This class is used to find files that were already snapshoted and clear out files that no longer referenced by any
  * snapshots
  */
-public class BlobStoreIndexShardSnapshots implements Iterable<BlobStoreIndexShardSnapshot> {
-    private final ImmutableList<BlobStoreIndexShardSnapshot> shardSnapshots;
+public class BlobStoreIndexShardSnapshots implements Iterable<SnapshotFiles>, ToXContent {
+    private final ImmutableList<SnapshotFiles> shardSnapshots;
+    private final ImmutableMap<String, FileInfo> files;
+    private final ImmutableMap<String, ImmutableList<FileInfo>> physicalFiles;
 
-    public BlobStoreIndexShardSnapshots(List<BlobStoreIndexShardSnapshot> shardSnapshots) {
+    public BlobStoreIndexShardSnapshots(List<SnapshotFiles> shardSnapshots) {
         this.shardSnapshots = ImmutableList.copyOf(shardSnapshots);
+        Map<String, FileInfo> files = newHashMap();
+        Map<String, List<FileInfo>> physicalFiles = newHashMap();
+        for (SnapshotFiles snapshot : shardSnapshots) {
+            for (FileInfo fileInfo : snapshot.indexFiles()) {
+                FileInfo oldFile = files.put(fileInfo.name(), fileInfo);
+                assert oldFile == null || oldFile.isSame(fileInfo);
+
+            }
+            for (FileInfo fileInfo : snapshot.indexFiles()) {
```

Comment:
Can you explain why we need to loop twice here? It seems like generating the `files` map and the `physicalFileList` could be done in the same loop?